src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/model/ISMRollup.kt

```diff
@@ -0,0 +1,199 @@
+        require(dimensions.filter { it.type == Dimension.Type.DATE_HISTOGRAM }.size == 1) {
+            "Must specify precisely one date histogram dimension"
+        }
+        require(dimensions.first().type == Dimension.Type.DATE_HISTOGRAM) { "The first dimension must be a date histogram" }
+        require(metrics.isNotEmpty()) { "Metrics cannot be empty" }
+    }
+
+    override fun toXContent(builder: XContentBuilder, params: ToXContent.Params): XContentBuilder {
+        builder.startObject()
+                .field(Rollup.DESCRIPTION_FIELD, description)
+                .field(Rollup.TARGET_INDEX_FIELD, targetIndex)
+                .field(Rollup.PAGE_SIZE_FIELD, pageSize)
+                .field(Rollup.DIMENSIONS_FIELD, dimensions)
+                .field(Rollup.METRICS_FIELD, metrics)
+        builder.endObject()
+        return builder
+    }
+
+    fun toRollup(sourceIndex: String, roles: List<String> = listOf()): Rollup {
+        val id = sourceIndex + toString()
+        val currentTime = Instant.now()
+        return Rollup(
+                id = DigestUtils.sha1Hex(id),
+                seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO,
+                primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM,
+                enabled = true,
+                schemaVersion = IndexUtils.DEFAULT_SCHEMA_VERSION,
+                jobSchedule = IntervalSchedule(Instant.now(), 1, ChronoUnit.MINUTES),
+                jobLastUpdatedTime = currentTime,
+                jobEnabledTime = currentTime,
+                description = this.description,
+                sourceIndex = sourceIndex,
+                targetIndex = this.targetIndex,
+                metadataID = null,
+                roles = roles,
+                pageSize = pageSize,
+                delay = null,
+                continuous = false,
+                dimensions = dimensions,
+                metrics = metrics
+        )
+    }
+
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : this(
+            description = sin.readString(),
+            targetIndex = sin.readString(),
+            pageSize = sin.readInt(),
+            dimensions = sin.let {
+                val dimensionsList = mutableListOf<Dimension>()
+                val size = it.readVInt()
+                for (i in 0 until size) {
+                    val type = it.readEnum(Dimension.Type::class.java)
+                    dimensionsList.add(
+                            when (requireNotNull(type) { "Dimension type cannot be null" }) {
+                                Dimension.Type.DATE_HISTOGRAM -> DateHistogram(sin)
+                                Dimension.Type.TERMS -> Terms(sin)
+                                Dimension.Type.HISTOGRAM -> Histogram(sin)
+                            }
+                    )
+                }
+                dimensionsList.toList()
+            },
+            metrics = sin.readList(::RollupMetrics)
+    )
+
+    override fun toString(): String {
```

Comment:
I believe the doc id has a limit which this could hit?

Reply:
Ah I see the DigestUtils above, ignore this

Reply:
I am doing a fixed length hash - sha1